### ISSUE TITLE
feat(forge): whitelist certain files to use ffi

### ISSUE THREAD

**Issue by OliverNChalk** [foundry-rs/foundry#2326]
### Component

Forge

### Describe the feature you would like

Currently, one can only enable/disable ffi for a given execution environment. This means that all running source code has access to `ffi` or none of it does. This means malicious libraries can do the following:

```solidity
try vm.ffi(malicious_task) {
    // running with ffi, do any additional tasks needed
} catch {
    // not running with ffi
}
```

One could argue that the user is required to understand & trust all their dependencies. But this seems slightly intense and could reasonably be missed by many users.

## Potential Solutions

To help aid managing the safety issue of `ffi`, one could imagine declaring the following property in `foundry.toml`:
```toml
ffi_enabled = [
    "test/__helpers/YulDeployer.sol",
    "test/__helpers/HuffDeployer.sol",
    "lib/forge-vyper/src/VyperDeployer.sol",
]
```

This would enable whitelisting a single file within a repository, greatly reducing the number of SLOC a user has to audit to use `ffi`.

## Considerations

- Unsure of the priority here given there has been a general goal to minimize required `ffi` usage through additional cheatcodes
- Unsure if it is easy to determine the calling `file`/`contract` for the purposes of vetting `ffi`

**Comment by gakonst** [comment]
Hmm this seems reasonable as an ask, but unsure how feasible it is, especially given inheritance etc.

**Comment by onbjerg** [comment]
There's two ways to go about this (only really one way):

- In the first case, we'd need to identify contracts as we execute the tests. We used to do this, but it's pretty expensive and requires a bigger refactor, so probably a non-starter
- We could do some sort of static analysis where we would check if any non-whitelisted files call FFI and exit if they do. This is more feasible but probably poor UX? Although if you don't whitelist the file, and it wanted to do FFI, I guess we'd have to abort execution in the first case as well, so it might be ok

**Comment by mds1** [comment]
> This would enable whitelisting a single file within a repository, greatly reducing the number of SLOC a user has to audit to use ffi.

Hmm, I wonder how much value this feature really even adds? If you just search the codebase for all calls to ffi with something like `grep -r \\.ffi`, that also greatly reduces the number of SLOC a user has to audit to use ffi 😅

Maybe a better solution is to add a command like `forge audit` (name based off [`yarn audit`](https://classic.yarnpkg.com/lang/en/docs/cli/audit/), maybe cargo has a similar command name we should use instead since "audit" already has a lot of connotations in the blockchain world) which basically executes that grep search and prints all uses of potentially dangerous cheatcodes like ffi or file reading/writing.

**Comment by gakonst** [comment]
i like that, there's [cargo audit](https://crates.io/crates/cargo-audit) which checks the lockfile for any crates with vulnerable editions. i can see that e.g. for checking if you have installed a bad Solmate version (mabye there was a bug in a release for example), or for printing you the `vm.ffi` calls

**Comment by OliverNChalk** [comment]
I like the audit solution, especially if it can leverage remappings to exhaustively check the referenced source code. Happy if we want to close this issue in favour of a new one, or re-write this one?

**Comment by onbjerg** [comment]
Yeah I think a dedicated `forge audit` issue makes sense where we can brainstorm more ideas :smile:

**Comment by OliverNChalk** [comment]
Closed in favour of #2360
